services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,342 @@
+        RoleName: roleName,
+      }),
+    });
+
+    // ROLE + POLICY
+    await readModifySet({
+      set: async () => await this.iam.attachRolePolicy({
+        PolicyArn: freshPolicyMetadata.Arn,
+        RoleName: roleName,
+      }),
+      read: async () => {
+        const entities = await this.iam.listEntitiesForPolicy({PolicyArn: freshPolicyMetadata.Arn});
+        const role = entities.PolicyRoles.find(pr => pr.RoleName === roleName);
+
+        if (role) {
+          // so this function does two things instead of 1
+          // also we always detach policy - even if the policy hasn't changed
+          // todo: refactor
+          await this.iam.detachRolePolicy({
+            PolicyArn: freshPolicyMetadata.Arn,
+            RoleName: role.RoleName,
+          });
+
+          return undefined;
+        }
+
+        return role;
+      },
+    });
+
+    // INSTANCE PROFILE
+    this.instanceProfile = await readModifySet({
+      read: async () => await this.iam.getInstanceProfile({InstanceProfileName: roleName}),
+      set: async () => await this.iam.createInstanceProfile({InstanceProfileName: roleName}),
+    });
+
+    //todo: do something to enable logging to stackdriver
+  }
+
+  async provision({workerPool}) {
+    const {workerPoolId} = workerPool;
+
+    if (!workerPool.providerData[this.providerId] || workerPool.providerData[this.providerId].running === undefined) {
+      await workerPool.modify(wt => {
+        wt.providerData[this.providerId] = wt.providerData[this.providerId] || {};
+        wt.providerData[this.providerId].running = wt.providerData[this.providerId].running || 0; // why zero?
+      });
+    }
+
+    const toSpawn = await this.estimator.simple({
+      workerPoolId,
+      ...workerPool.config,
+      running: workerPool.providerData[this.providerId].running,
+    });
+
+    let spawned;
+
+    try {
+      spawned = await this.ec2.runInstances({
+        IamInstanceProfile: {
+          Arn: this.instanceProfile.Arn,
+          Name: this.instanceProfile.InstanceProfileName,
+        },
+        MaxCount: toSpawn,
+        MinCount: toSpawn,
+        ImageId: workerPool.config.imageId,
+        InstanceType: workerPool.config.instanceType,
+        Placement: {
+          AvailabilityZone: this.region,
+        },
+        TagSpecifications: {
+          ResourceType: 'instance',
+          Tags: [{
+            Key: 'Provider',
+            Value: `wm-${this.providerId}`,
+          }, {
+            Key: 'Owner',
+            Value: workerPool.owner,
+          }, {
+            Key: 'Project',
+            Value: this.project,
+          }],
+        },
+      });
+    } catch (e) {
+      await workerPool.reportError({
+        kind: 'creation-error',
+        title: 'Instance Creation Error',
+        description: e.message, // TODO: Make sure we clear exposing this with security folks
+        notify: this.notify,
+        WorkerPoolError: this.WorkerPoolError,
+      });
```

Comment:
oh I thought `workerPool.reportError` was a side effect function so returning it would make no difference, and the method would return anyways upon execution of the `catch` block so no reason to put `return` in there. I can add the keyword tho, sure